File: 2022/day13.hs
-- Advent of Code 2022, day 13
--  Venanzio Capretta

module Main where

import System.Environment
import Data.List
import Data.Char

import Control.Applicative
import qualified Data.Map as M

import FunParser
import AoCTools

main :: IO ()
main = do
  args <- getArgs
  puzzle (head args)

puzzle :: String -> IO ()
puzzle fileName = do
  input <- readFile fileName
  let xs = parseAll pInput input
  putStrLn ("Part 1: " ++ show (part1 xs))
  putStrLn ("Part 2: " ++ show (part2 xs))

-- Parsing the input

data Packet = PValue Int | PList [Packet]
  deriving (Show,Eq)

pPacket :: Parser Packet
pPacket = do n <- natural
             return (PValue n)
          <|>
          do ps <- delim "[" (manySepStr pPacket ",") "]"
             return (PList ps)

pInput :: Parser [(Packet,Packet)]
pInput = many (do p1 <- pPacket
                  p2 <- pPacket
                  return (p1,p2))

-- Part 1

instance Ord Packet where
  -- compare :: Packet -> Packet -> Ordering
  compare (PValue n1)   (PValue n2)   = compare n1 n2
  compare (PList ps1)   (PList ps2)   = compare ps1 ps2
  compare p1@(PValue _) p2            = compare (PList [p1]) p2
  compare p1            p2@(PValue _) = compare p1 (PList [p2])

part1 :: [(Packet,Packet)] -> Int
part1 ps = sum $ map ((+1).fst) $ filterIndices (\(p1,p2) -> p1<=p2) ps

-- Part 2

divP1 :: Packet 
divP1 = PList [PList [PValue 2]]

divP2 :: Packet 
divP2 = PList [PList [PValue 6]]
  
flatten :: [(Packet,Packet)] -> [Packet]
flatten [] = []
flatten ((p1,p2):ps) = p1 : p2 : flatten ps

decoderKey :: [Packet] -> Int
decoderKey ps =
  let Just i1 = elemIndex divP1 ps
      Just i2 = elemIndex divP2 ps
  in (i1+1) * (i2+1)

part2 :: [(Packet,Packet)] -> Int
part2 = decoderKey . sort . ([divP1,divP2]++) . flatten

File: 2022/python/day13.py
# Advent of Code 2022, day 13
#   Venanzio Capretta

import parsers as P
import functools as fun

print("Advent of Code 2022, Day 13")

f = open("../input13")
input = [P.Source(l) for l in f.read().splitlines()]
f.close()

# Parsing the input

def parse_packet(s):
  x = P.num(s)
  if x != None:
    return x
  elif s.next() == '[':
    P.word('[',s)
    l = P.lst(parse_packet,s)
    P.word(']',s)
    return l
  else:
    return None

packets = [parse_packet(l) for l in input if l.text!='']

# Part 1

def lexicographic(ord,l1,l2):
  if l1 == []:
    return True
  elif l2 == []:
    return False
  elif l1[0] == l2[0]:
    return lexicographic(ord,l1[1:],l2[1:])
  else:
    return ord(l1[0],l2[0])

def packet_ord(p1,p2):
  if type(p1)==type(p2):
    if type(p1) == type(0):
      return p1<p2
    else:
      return lexicographic(packet_ord,p1,p2)
  else:
    if type(p1) == type(0):
      return packet_ord([p1],p2)
    else:
      return packet_ord(p1,[p2])


s=0
for i in range(0,len(packets),2):
  if packet_ord(packets[i],packets[i+1]):
    s += i//2+1

print("Part 1: ")
print(s)

# Part 2

packets.append([[2]])
packets.append([[6]])

def compare(p1,p2):
  if packet_ord(p1,p2):
    return -1
  else:
    return 1

packets.sort(key=fun.cmp_to_key(compare))

print("Part 2: ")
print((packets.index([[2]])+1) * (packets.index([[6]])+1))
